pkg/backtest/exchange.go

```diff
@@ -362,21 +362,36 @@ func (e *Exchange) SubscribeMarketData(startTime, endTime time.Time, extraInterv
 	return klineC, nil
 }
 
-func (e *Exchange) ConsumeKLine(k types.KLine) {
-	if k.Interval == types.Interval1m {
-		e.currentTime = k.EndTime.Time()
-
-		matching, ok := e.matchingBook(k.Symbol)
-		if !ok {
-			log.Errorf("matching book of %s is not initialized", k.Symbol)
-			return
+func (e *Exchange) ConsumeKLine(k types.KLine, handlers []func(types.KLine, *ExchangeDataSource), src *ExchangeDataSource) {
+	matching, ok := e.matchingBook(k.Symbol)
+	if !ok {
+		log.Errorf("matching book of %s is not initialized", k.Symbol)
+		return
+	}
+	if matching.ParamCache == nil {
+		matching.ParamCache = make(map[types.Interval]Param)
+	}
+	_, ok = matching.ParamCache[k.Interval]
+	if ok { // pop out all the old
+		for _, param := range matching.ParamCache {
+			if param.kline.Interval == types.Interval1m {
+				// here we generate trades and order updates
+				matching.processKLine(param.kline)
+				matching.NextKLine = &k
+			}
+			// log.Errorf("kline %v, next %v", param.kline, matching.NextKLine)
+			e.MarketDataStream.EmitKLineClosed(param.kline)
+			for _, h := range param.callbacks {
+				h(param.kline, param.src)
+			}
 		}
-
-		// here we generate trades and order updates
-		matching.processKLine(k)
+		matching.ParamCache = make(map[types.Interval]Param)
+	}
+	matching.ParamCache[k.Interval] = Param{
+		callbacks: handlers,
+		src:       src,
+		kline:     k,
```

Comment:
there will be a delay for intervals like 1h, 2h, 4h

1h kline closed event will be delayed for 1h
2h will be delayed for 2h
4h will be delayed for 4h